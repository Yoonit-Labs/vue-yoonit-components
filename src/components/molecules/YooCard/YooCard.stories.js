import YooCard from './YooCard.vue'
import PropsConfig from './YooCard.config'

export default {
  title: 'Yoonit/YooCard',
  component: YooCard,
  argTypes: {
    justifySlotContent: {
      control: {
        type: 'select',
        options: PropsConfig.justifySlotContent.options
      }
    },
    borderFill: {
      control: {
        type: 'select',
        options: PropsConfig.borderFill.options
      }
    }
  }
}

const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { YooCard },
  template:
    '<yoo-card v-bind="$props"></yoo-card>'
})

export const Custom = Template.bind({})
Custom.args = { title: 'YooCard', subtitle: 'Test' }

export const Neutral = Template.bind({})
Neutral.args = { title: 'YooCard', subtitle: 'Test', borderFill: 'neutral' }

export const Primary = Template.bind({})
Primary.args = { title: 'YooCard', subtitle: 'Test', borderFill: 'primary' }

export const Danger = Template.bind({})
Danger.args = { title: 'YooCard', subtitle: 'Test', borderFill: 'danger' }

export const Light = Template.bind({})
Light.args = { title: 'YooCard', subtitle: 'Test', borderFill: 'light' }

export const Dark = Template.bind({})
Dark.args = { title: 'YooCard', subtitle: 'Test', borderFill: 'dark' }

export const Darkest = Template.bind({})
Darkest.args = { title: 'YooCard', subtitle: 'Test', borderFill: 'darkest' }