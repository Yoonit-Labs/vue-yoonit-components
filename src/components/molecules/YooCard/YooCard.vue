<template lang="pug">
  YooGridLayout(
    :class="['yoo-card', ...takeModifier]"
    rows="auto"
    cols="*, auto"
    width="100%"
  )
    YooFlexLayout(
      flexDirection="column"
      col="1"
      row="1"
      flexWrap="wrap"
      justifyContent="center"
      :class="['yoo-card__item', ...takeItemsModifier]"
    )
      h3(
        class="yoo-card__title"
      ) {{title}}

      h3(
        class="yoo-card__title"
        v-show="false"
      ) {{privateKey}}

      p(
        v-show="subtitle",
        class="yoo-card__subtitle"
        width="100%"
      ) {{subtitle}}

    YooFlexLayout(
      v-if="controlVisibility"
      col="2"
      row="1"
      :justifyContent="justifySlotContent"
      alignItems="center"
      gap="8px"
    )
      slot(
        name="control"
      )
</template>

<script>

import YooFlexLayout from '@/components/quarks/YooFlexLayout/YooFlexLayout.vue'
import YooGridLayout from '@/components/quarks/YooGridLayout/YooGridLayout.vue'

export default {
  name: 'YooCard',
  props: {
    title: {
      type: String,
      required: false,
      default: ''
    },
    subtitle: {
      type: String,
      required: false,
      default: ''
    },
    justifySlotContent: {
      type: String,
      required: false,
      default: 'center',
      validator: value => [
        'flex-start',
        'flex-end',
        'center',
        'space-between',
        'space-around'
      ]
        .includes(value)
    },
    privateKey: {
      type: [Date, String, Number],
      required: false,
      default: Date.now()
    },
    borderLeft: {
      type: Boolean,
      default: true
    },
    borderFill: {
      type: String,
      required: false,
      default: 'neutral',
      validator: value => [
        'neutral',
        'primary',
        'danger',
        'light',
        'dark',
        'darkest'
      ]
        .includes(value)
    }
  },
  components: {
    YooGridLayout,
    YooFlexLayout
  },
  data: () => ({}),
  beforeCreate () {},
  created () {},
  beforeMount () {},
  mounted () {},
  beforeUpdate () {},
  updated () {},
  beforeDestroy () {},
  destroyed () {},
  computed: {
    takeModifier () {
      return ''
    },
    takeItemsModifier () {
      const blockItem = 'yoo-card__item'
      const classList = []

      if (this.borderLeft) {
        classList
          .push(`${blockItem}--border-left`)
      }

      classList
        .push(`${blockItem}--border-fill-${this.borderFill}`)

      return classList
    },
    controlVisibility () {
      return !!this.$slots.control
    }
  },
  methods: {},
  filters: {},
  watch: {}
}
</script>

<style src="./YooCard.sass" lang="sass" scoped></style>
