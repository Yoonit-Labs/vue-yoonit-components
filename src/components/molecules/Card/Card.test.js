import { shallowMount } from '@vue/test-utils'
import YooCard from '@/components/molecules/Card/Card.vue'
import PropsConfig from './Card.config'

const classBlock = 'yoo-card'
const SlotText = 'Default Slot Text'

const mountComponent = () => {
  return shallowMount(YooCard, {
    slots: { default: SlotText },
    propsData: {
      title: 'Default Title',
      subtitle: 'Default Subtitle'
    }
  })
}

describe('YooCard Component', () => {
  let wrapper
  beforeEach(() => {
    wrapper = mountComponent()
  })

  it('Matches Snapshot', async () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Loads the Component HTML', async () => {
    expect(wrapper.classes('yoo-card')).toBe(true)
    expect(wrapper.find('.yoo-card__title').exists()).toBe(true)
    expect(wrapper.find('.yoo-card__subtitle').exists()).toBe(true)
  })

  describe('Props', () => {
    describe('borderLeft', () => {
      it('Has a valid default value', () => {
        expect(YooCard.props.borderLeft.default).toBe(true)
      })

      it(`Includes borderLeft class: .${classBlock}__style--border-left`, async () => {
        expect(wrapper.find(`.${classBlock}__style--border-left`).exists()).toBe(true)
      })

      it(`Does not includes borderLeft class: .${classBlock}__style--border-left`, async () => {
        await wrapper.setProps({ borderLeft: false })
        expect(wrapper.find(`.${classBlock}__style--border-left`).exists()).toBe(false)
      })
    })

    describe('borderFill', () => {
      it('Has a valid default value', () => {
        expect(PropsConfig.borderFill.options.includes(YooCard.props.borderFill.default)).toBe(true)
      })
      PropsConfig.borderFill.options.forEach(borderFill => {
        it(`Includes borderFill class: .${classBlock}__fill--border-${borderFill}`, async () => {
          await wrapper.setProps({ borderFill })
          expect(wrapper.find(`.${classBlock}__fill--border-${borderFill}`).exists()).toBe(true)
        })
      })
    })
  }) // describe Props
})
