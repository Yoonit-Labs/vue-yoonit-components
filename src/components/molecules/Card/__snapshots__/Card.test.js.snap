// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooCard Component Matches Snapshot 1`] = `
<yoogridlayout-stub rows="auto" cols="*, auto" gap="0" width="100%" horizontalalign="stretch" verticalalign="stretch" class="yoo-card">
  <yooflexlayout-stub flexdirection="column" flexwrap="wrap" justifycontent="center" alignitems="stretch" aligncontent="stretch" gap="0" col="1" row="1" class="m__t--l m__b--l p__t--x p__b--x p__l--xl yoo-card__item yoo-card__item--border-left yoo-card__item--border-fill-neutral">
    <h3 class="yoo-card__title">Default Title</h3>
    <p width="100%" class="yoo-card__subtitle">Default Subtitle</p>
  </yooflexlayout-stub>
  <!---->
</yoogridlayout-stub>
`;
